pkg/typesystem/typesystem_test.go

```diff
@@ -960,6 +960,54 @@ func TestInvalidRelationTypeRestrictionsValidations(t *testing.T) {
 			},
 			err: &InvalidRelationError{ObjectType: "document", Relation: "parent"},
 		},
+		{
+			name: "WildcardNotAllowedInTheTuplesetPartOfTTU",
+			model: &openfgapb.AuthorizationModel{
+				SchemaVersion: SchemaVersion1_1,
+				TypeDefinitions: []*openfgapb.TypeDefinition{
+					{
+						Type: "user",
+					},
+					{
+						Type: "folder",
+						Relations: map[string]*openfgapb.Userset{
+							"viewer": This(),
+						},
+						Metadata: &openfgapb.Metadata{
+							Relations: map[string]*openfgapb.RelationMetadata{
+								"viewer": {
+									DirectlyRelatedUserTypes: []*openfgapb.RelationReference{
+										DirectRelationReference("user", ""),
+									},
+								},
+							},
+						},
+					},
+					{
+						Type: "document",
+						Relations: map[string]*openfgapb.Userset{
+							"parent": This(),
+							"viewer": Union(This(), TupleToUserset("parent", "viewer")),
+						},
+						Metadata: &openfgapb.Metadata{
+							Relations: map[string]*openfgapb.RelationMetadata{
+								"parent": {
+									DirectlyRelatedUserTypes: []*openfgapb.RelationReference{
+										WildcardRelationReference("folder"),
+									},
+								},
+								"viewer": {
+									DirectlyRelatedUserTypes: []*openfgapb.RelationReference{
+										DirectRelationReference("user", ""),
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			err: &InvalidRelationError{ObjectType: "document", Relation: "parent"},
```

Comment:
Should we add more context to this error so that it is more clear what the violation was due to?

For example, it'd be nice to have an error message along the lines of
> 'folder:*' is an invalid type restriction for relation 'document#parent' because it is defined under a tupleset relation.

Reply:
Yeah. I am happy to have better errors. What I am not so happy about is trying to "match" them in tests. Should we refactor the error to be something like
```
type InvalidRelationError struct {
	Reason string
}
```
and fill in the reason on a case-by-case basis? Or have a general error wrapping an
```
var InvalidRelationError = errors.New("invalid relation")
```
which we can add context to, and also match against in tests?

Reply:
I think until we rework our errors it would be more consistent with the existing code to have a custom error type `InvalidRelationError` with the underlying message (your first example). Then just implement the `Is(target interface{})` method and we can match in tests using `require.ErrorIs(...)`